README.md
# Arduino-Notes
This is a library for the Arduino Uno to play tones a lot easier

To learn more about it visit https://domi04151309.github.io/Arduino/Notes


src/CHANGELOG.txt
-Removed compiler error
-Added better tone detection

-Added ability to turn off tone LEDs
-Added switch case instead of if else

-Added Errors with board detection
-Performance improvements
-Added board detection

-Changed function beatLEDFunctionFour() to beatLEDFunctionFour(int)
-Added example GermanDuckSongShortPitches
-Added definitions for the notes in pitches.h

-Added example GermanDuckSongShort
-Added ability to shorten some functions
-Added Errors in serial monitor

-Added example GermanDuckSong
-Added serial monitor for debugging
-Changed some public integers to privat integers
-Added comments for description

-Fixed Bugs
-Added everything

src/Board.h
/*
 *Notes - Library for a better tone command
 *
 *Copyright 2018 Dominik Reichl
 *
 *Developed by Dominik Reichl
 *
 *This library is free software; you can redistribute it and/or modify
 *it under the terms of the GNU General Public License as published by
 *the Free Software Foundation; either version 2 of the License, or
 *(at your option) any later version.
 *
 *This library is distributed in the hope that it will be useful,
 *but WITHOUT ANY WARRANTY; without even the implied warranty of
 *MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *GNU General Public License for more details.
 *
 *You should have received a copy of the GNU General Public License
 *along with this library; if not, write to the Free Software
 *Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 *
 *As a special exception, you may use this file as part of a free software
 *library without restriction.  Specifically, if other files instantiate
 *templates or use macros or inline functions from this file, or you compile
 *this file and link it with other files to produce an executable, this
 *file does not by itself cause the resulting executable to be covered by
 *the GNU General Public License.  This exception does not however
 *invalidate any other reasons why the executable file might be covered by
 *the GNU General Public License.
 */

#ifndef Board_h
#define Board_h
 
#if defined(__AVR_ATmega328P__)
	#define ARDUINO_BOARD_UNO
	#define ARDUINO_BOARD_FIO
	
#elif defined(__AVR_ATSAMD21G18A__)
	#define ARDUINO_BOARD_ZERO
	
#elif defined(__AVR_ATSAM3X8E__)
	#define ARDUINO_BOARD_DUE
	
#elif defined(__AVR_Atmega32U4__)
	#define ARDUINO_BOARD_YUN_16MHZ
	#define ARDUINO_BOARD_MICRO
	#define ARDUINO_BOARD_LEONARDO
	#define ARDUINO_BOARD_ESPLORA
	
#elif defined(_AVR_AR9331__)
	#define ARDUINO_BOARD_YUN_400MHZ
	
#elif defined(__AVR_ATmega16U4__)
	#define ARDUINO_BOARD_LEONARDO
	
#elif defined(__AVR_ATmega1280__) || defined(__AVR_ATmega2560__)
	#define ARDUINO_BOARD_MEGA_NORMAL
	#define ARDUINO_BOARD_MEGA_ADK
	
#elif defined(_AVR_ATmega328__)
	#define ARDUINO_BOARD_NANO
	#define ARDUINO_BOARD_PRO_NORMAL
	#define ARDUINO_BOARD_PRO_328
	#define ARDUINO_BOARD_ETHERNET
	
#elif defined(_AVR_ATmega168__)
	#define ARDUINO_BOARD_NANO
	#define ARDUINO_BOARD_UNO
	#define ARDUINO_BOARD_PRO_NORMAL
	
#elif defined(_AVR_ATmega168V__)
	#define ARDUINO_BOARD_LILYPAD_1
	
#elif defined(_AVR_ATmega328V__)
	#define ARDUINO_BOARD_LILYPAD_2
	
#elif defined(_AVR_ATTiny85__)
	#define ARDUINO_BOARD_TRINKET
	
#elif  defined(__AVR_ARCv2EM__) || (__CURIE_FACTORY_DATA_H_)
	#define ARDUINO_BOARD_INTEL_CURIE
	#define ARDUINO_BOARD_101
	
#else
	#define ARDUINO_BOARD_UNKNOWN

#endif

#endif

src/Pitches.h
/*************************************************
 * Public Constants
 *************************************************/
 
#ifndef Pitches_h
#define Pitches_h
 
#define NOTE_C1  33
#define NOTE_CS1 34.8
#define NOTE_D1  37.1
#define NOTE_DS1 39.6
#define NOTE_E1  41.25
#define NOTE_F1  44
#define NOTE_FS1 46.4
#define NOTE_G1  49.5
#define NOTE_GS1 52.8
#define NOTE_A1  55
#define NOTE_AS1 59.4
#define NOTE_B1  61.85

#define NOTE_C2  66
#define NOTE_CS2 69.6
#define NOTE_D2  74.2
#define NOTE_DS2 79.2
#define NOTE_E2  82.5
#define NOTE_F2  88
#define NOTE_FS2 92.8
#define NOTE_G2  99
#define NOTE_GS2 105.6
#define NOTE_A2  110
#define NOTE_AS2 118.8
#define NOTE_B2  123.7

#define NOTE_C3  132
#define NOTE_CS3 139.2
#define NOTE_D3  148.4
#define NOTE_DS3 158.4
#define NOTE_E3  165
#define NOTE_F3  176
#define NOTE_FS3 185.6
#define NOTE_G3  198
#define NOTE_GS3 211.2
#define NOTE_A3  220
#define NOTE_AS3 237.6
#define NOTE_B3  247.4

#define NOTE_C4  264
#define NOTE_CS4 278.4
#define NOTE_D4  296.8
#define NOTE_DS4 316.8
#define NOTE_E4  330
#define NOTE_F4  352
#define NOTE_FS4 371.2
#define NOTE_G4  396
#define NOTE_GS4 422.4
#define NOTE_A4  440
#define NOTE_AS4 475.2
#define NOTE_B4  494.8

#define NOTE_C5  528
#define NOTE_CS5 556.8
#define NOTE_D5  593.6
#define NOTE_DS5 633.6
#define NOTE_E5  660
#define NOTE_F5  704
#define NOTE_FS5 742.4
#define NOTE_G5  792
#define NOTE_GS5 844.8
#define NOTE_A5  880
#define NOTE_AS5 950.4
#define NOTE_B5  989.6

#define NOTE_C6  1056
#define NOTE_CS6 1113.6
#define NOTE_D6  1187.2
#define NOTE_DS6 1267.2
#define NOTE_E6  1320
#define NOTE_F6  1408
#define NOTE_FS6 1484.8
#define NOTE_G6  1584
#define NOTE_GS6 1689.6
#define NOTE_A6  1760
#define NOTE_AS6 1900.8
#define NOTE_B6  1979.2

#define NOTE_C7  2112
#define NOTE_CS7 2227.2
#define NOTE_D7  2374.4
#define NOTE_DS7 2534.4
#define NOTE_E7  2640
#define NOTE_F7  2816
#define NOTE_FS7 2969.6
#define NOTE_G7  3168
#define NOTE_GS7 3379.2
#define NOTE_A7  3520
#define NOTE_AS7 3801.6
#define NOTE_B7  3958.4

#endif

Notes/src/Notes.h
/*
 *Notes - Library for a better tone command
 *
 *Copyright 2018 Dominik Reichl
 *
 *Developed by Dominik Reichl
 *
 *This library is free software; you can redistribute it and/or modify
 *it under the terms of the GNU General Public License as published by
 *the Free Software Foundation; either version 2 of the License, or
 *(at your option) any later version.
 *
 *This library is distributed in the hope that it will be useful,
 *but WITHOUT ANY WARRANTY; without even the implied warranty of
 *MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *GNU General Public License for more details.
 *
 *You should have received a copy of the GNU General Public License
 *along with this library; if not, write to the Free Software
 *Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 *
 *As a special exception, you may use this file as part of a free software
 *library without restriction.  Specifically, if other files instantiate
 *templates or use macros or inline functions from this file, or you compile
 *this file and link it with other files to produce an executable, this
 *file does not by itself cause the resulting executable to be covered by
 *the GNU General Public License.  This exception does not however
 *invalidate any other reasons why the executable file might be covered by
 *the GNU General Public License.
 */
 
#include <Pitches.h>
#ifndef Notes_h
#define Notes_h

#define ON true
#define OFF false

class Notes
{
 public:
   Notes(unsigned int x, unsigned int y);
   void speakerSetup(uint8_t speaker, unsigned int speaker2 = 120);
   void toneLEDSetup(uint8_t pin0, uint8_t pin1, uint8_t pin2, uint8_t pin3, uint8_t pin4, uint8_t pin5, uint8_t pin6, uint8_t pin7);
	 void toneLEDState(bool state = true);
   void blinkLED(uint8_t param1LED);
   void note(float param1, uint8_t param2, bool param3 = false, bool param4 = false);
 private:
	 bool toneLEDStateBool = false;
   uint8_t speakerPort;
   uint8_t beatLEDPort;
   uint8_t cLED;
   uint8_t dLED;
   uint8_t eLED;
   uint8_t fLED;
   uint8_t gLED;
   uint8_t aLED;
   uint8_t bLED;
   uint8_t toneLED;
   unsigned int bpm;
   float beatDuration;
   unsigned int beatLEDTime;
   unsigned int toneDelay;
   void beatLEDFunction();
   void beatLEDFunctionFour(uint8_t z = 1);
   void noToneDelay();
};

#endif

Notes/src/Notes.cpp
/*
 *Notes - Library for a better tone command
 *
 *Copyright 2018 Dominik Reichl
 *
 *Developed by Dominik Reichl
 *
 *This library is free software; you can redistribute it and/or modify
 *it under the terms of the GNU General Public License as published by
 *the Free Software Foundation; either version 2 of the License, or
 *(at your option) any later version.
 *
 *This library is distributed in the hope that it will be useful,
 *but WITHOUT ANY WARRANTY; without even the implied warranty of
 *MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *GNU General Public License for more details.
 *
 *You should have received a copy of the GNU General Public License
 *along with this library; if not, write to the Free Software
 *Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 *
 *As a special exception, you may use this file as part of a free software
 *library without restriction.  Specifically, if other files instantiate
 *templates or use macros or inline functions from this file, or you compile
 *this file and link it with other files to produce an executable, this
 *file does not by itself cause the resulting executable to be covered by
 *the GNU General Public License.  This exception does not however
 *invalidate any other reasons why the executable file might be covered by
 *the GNU General Public License.
 */

#include <Arduino.h>
#include <Notes.h>
#include <Board.h>

//Setup process
Notes::Notes(unsigned int x, unsigned int y){ 
  beatLEDTime=x;
  toneDelay=y;
}

//Sets the speakerPort and the speed
void Notes::speakerSetup(uint8_t speaker, unsigned int speed){
  //For Debugging
  Serial.print("speakerSetup ");Serial.print(speaker);Serial.print(", ");Serial.println(speed);
  
  #ifdef ARDUINO_BOARD_UNO
  if(speaker != 3 && speaker != 5 && speaker != 6 && speaker != 9 && speaker != 10 && speaker != 11){
		Serial.println("###");
		Serial.println("### ERROR: Pin has to be 3, 5, 6, 9, 10 or 11");
		Serial.println("###");
  }
  #elif defined ARDUINO_BOARD_ZERO
  if(speaker != 3 && speaker != 4 && speaker != 5 && speaker != 6 && speaker != 8 && speaker != 9 && speaker != 10 && speaker != 11 && speaker != 12 && speaker != 13){
		Serial.println("###");
		Serial.println("### ERROR: Pin has to be 3, 4, 5, 6, 8, 9, 10, 11, 12 or 13");
		Serial.println("###");
  }
  #elif defined ARDUINO_BOARD_YUN_400MHZ
  if(speaker != 3 && speaker != 5 && speaker != 6 && speaker != 10 && speaker != 11){
		Serial.println("###");
		Serial.println("### ERROR: Pin has to be 3, 5, 6, 10 or 11");
		Serial.println("###");
  }
  #endif
  
  speakerPort=speaker;
  bpm=speed;
  beatDuration=60000/bpm;
}

//Sets the ports of the LED pins
void Notes::toneLEDSetup(uint8_t pin0, uint8_t pin1, uint8_t pin2, uint8_t pin3, uint8_t pin4, uint8_t pin5, uint8_t pin6, uint8_t pin7){
  //For Debugging
  Serial.print("toneLEDSetup ");Serial.print(pin0);Serial.print(", ");Serial.print(pin1);Serial.print(", ");Serial.print(pin2);Serial.print(", ");Serial.print(pin3);Serial.print(", ");Serial.print(pin4);Serial.print(", ");Serial.print(pin5);Serial.print(", ");Serial.print(pin6);Serial.print(", ");Serial.println(pin7);
  
  #if defined(ARDUINO_BOARD_UNO) || defined(ARDUINO_BOARD_ZERO) || defined(ARDUINO_BOARD_YUN_400MHZ)
  if(pin0 > 14 || pin1 > 14 || pin2 > 14 || pin3 > 14 || pin4 > 14 || pin5 > 14 || pin6 > 14 || pin7 > 14 || pin0 < 2 || pin1 < 2 || pin2 < 2 || pin3 < 2 || pin4 < 2 || pin5 < 2 || pin6 < 2 || pin7 < 2){
		Serial.println("###");
		Serial.println("### ERROR: Pin numbers have to be between 2 and 13");
		Serial.println("###");
  }
  #endif
  
  beatLEDPort=pin0;
  cLED=pin1;
  dLED=pin2;
  eLED=pin3;
  fLED=pin4;
  gLED=pin5;
  aLED=pin6;
  bLED=pin7;
  pinMode(beatLEDPort, OUTPUT);
  pinMode(cLED, OUTPUT);
  pinMode(dLED, OUTPUT);
  pinMode(eLED, OUTPUT);
  pinMode(fLED, OUTPUT);
  pinMode(gLED, OUTPUT);
  pinMode(aLED, OUTPUT);
  pinMode(bLED, OUTPUT);
}

void Notes::toneLEDState(bool state){
	toneLEDStateBool = state;
}

//A LED blinks
void Notes::blinkLED(uint8_t param1LED){
  //For Debugging
  Serial.print("blinkLED ");Serial.println(param1LED);
  digitalWrite(param1LED,HIGH);
  delay(beatDuration/4);
  digitalWrite(param1LED,LOW);
}

//BeatLED blinks
void Notes::beatLEDFunction(){
  digitalWrite(beatLEDPort,HIGH);
  delay(beatLEDTime);
  digitalWrite(beatLEDPort,LOW);
}

//BeatLED blink + 1/4 note
void Notes::beatLEDFunctionFour(uint8_t z){
  for(uint8_t repeats=z; repeats>0; repeats=repeats-1){
    beatLEDFunction();
    delay(beatDuration-beatLEDTime);
  }
}

//noTone with delay after every note in case of two equal notes after each other
void Notes::noToneDelay(){
  noTone(speakerPort);
  delay(toneDelay);
}

//Main function of this library
void Notes::note(float param1, uint8_t param2, bool param3, bool param4){
    
  //For Debugging
  Serial.print("note ");Serial.print(param1);Serial.print(", ");Serial.print(param2);Serial.print(", ");Serial.print(param3);Serial.print(", ");Serial.println(param4);
  
  //Plays the sound
  tone(speakerPort,param1);
  
	if(toneLEDStateBool == true){
		//LEDs for the notes
		int x;
		for(x=param1;x>NOTE_B1;x=x/2){
			Serial.println(x);
		}
		if(x<=NOTE_C1){
		  toneLED=cLED;
		}else if(x<=NOTE_D1){
		  toneLED=dLED;
		}else if(x<=NOTE_E1){
		  toneLED=eLED;
		}else if(x<=NOTE_F1){
		  toneLED=fLED;
		}else if(x<=NOTE_G1){
		  toneLED=gLED;
		}else if(x<=NOTE_A1){
		  toneLED=aLED;
		}else if(x<=NOTE_B1){
		  toneLED=bLED;
		}
		
		//Turns on the toneLED
		digitalWrite(toneLED,HIGH);
	}
  
  if(param4 == false){
		switch(param2){
			case 1:
				//1/1 note
				beatLEDFunctionFour(3);
				beatLEDFunction();
				delay(beatDuration-beatLEDTime-toneDelay);
				noToneDelay();
				break;
				
			case 2:
				if(param3 == true){
					//1/2 note with point
					beatLEDFunctionFour(2);
					beatLEDFunction();
					delay(beatDuration-beatLEDTime-toneDelay);
					noToneDelay();
				}else if(param3 == false){
					//1/2 note
					beatLEDFunctionFour();
					beatLEDFunction();
					delay(beatDuration-beatLEDTime-toneDelay);
					noToneDelay();
				}
				break;
				
			case 4:
				if(param3 == true){
					//1/4 note with point
					beatLEDFunctionFour();
					beatLEDFunction();
					delay(beatDuration/2-beatLEDTime-toneDelay);
					noToneDelay();
				}else if(param3 == false){
					//1/4 note
					beatLEDFunction();
					delay(beatDuration-beatLEDTime-toneDelay);
					noToneDelay();
				}
				break;
				
			case 8:
				//1/8 note
				beatLEDFunction();
				delay(beatDuration/2-beatLEDTime-toneDelay);
				noToneDelay();
				break;
				
			default:
				Serial.println("###");
				Serial.println("### ERROR: Parameter 2 has to be 1, 2, 4 or 8");
				Serial.println("###");
				break;
    }
	
  }else if(param4 == true){
		switch(param2){
			case 1:
				//1/1 note with asynchronous beatLED
				delay(beatDuration/2);
				beatLEDFunctionFour(3);
				beatLEDFunction();
				delay(beatDuration/2-beatLEDTime-toneDelay);
				noToneDelay();
				break;
				
			case 2:
				if(param3 == true){
				//1/2 note with asynchronous beatLED and point
					delay(beatDuration/2);
					beatLEDFunctionFour(2);
					beatLEDFunction();
					delay(beatDuration/2-beatLEDTime-toneDelay);
					noToneDelay();
				}else if(param3 == false){
				//1/2 note with asynchronous beatLED
					delay(beatDuration/2);
					beatLEDFunctionFour();
					beatLEDFunction();
					delay(beatDuration/2-beatLEDTime-toneDelay);
					noToneDelay();
				}
				break;
				
			case 4:
				if(param3 == true){
				//1/4 note with asynchronous beatLED and point
					delay(beatDuration/2);
					beatLEDFunction();
					delay(beatDuration-beatLEDTime-toneDelay);
					noToneDelay();
				}else if(param3 == false){
				//1/4 note with asynchronous beatLED
					delay(beatDuration/2);
					beatLEDFunction();
					delay(beatDuration/2-beatLEDTime-toneDelay);
					noToneDelay();
				}
				break;
				
			case 8:
				//1/8 note with asynchronous beatLED
				delay(beatDuration/2-toneDelay);
				noToneDelay();
				break;
				
			default:
				Serial.println("###");
				Serial.println("### ERROR: Parameter 2 has to be 1, 2, 4 or 8");
				Serial.println("###");
				break;
    }
  }	
	if(toneLEDStateBool == true){
		//Turns off the toneLED
		digitalWrite(toneLED,LOW);
	}
}